### DISCUSSION TITLE
How to use live update mechanism for own purposes

### DISCUSSION THREAD

**Discussion by wega789** [osmandapp/OsmAnd#15648] [Q&A]
I use OsmAnd+ and want to apply some local changes to a map: adjust some waypoints in cycleways and footpaths, create new ones or change some tags. These changes should not end up in the OSM database right now. 

My idea was to follow the approach suggested in #13543 (use of fake OsmAnd Live delta file), i.e. edit a small local area with JOSM, convert it to obf with OsmAndMapCreator and copy this file into the net.osmand.plus/files directory on my Galaxy A3. In the edited region it overlaps the basic map. I was aware of the filename convention mentioned in #7562. 

Unfortunately this only worked for creating new paths. Moved paths show up twice, in their old position (from the basic map) and their new position (from the edited map). I checked that the id's of ways and nodes are not changed when edited with JOSM. According to the previous discussions in case of duplicate id's the file sort order should determine which object is rendered. However, using the -vmap -vmapobjects paramter for inspector I noticed that the obf file somehow uses two ids, the osm-id and an id of its own. The first one is identical in both obf files, but the second is different (I dont know how to check for nodes). 

What am I doing wrong or is it possible at all?

**Comment by vshcherb** [comment]
Yes, if you have 2 files "old" & "new", you can generate 2 obf files and then use utility from OsmAndMapCreator to generate "diff" file, once you have generated diff file you can put it in OsmAnd files and it should work

**Reply by wega789** [reply]
Thank you for your assistance. However, I need some more help:

"... if you have 2 files "old" & "new" ...":
So far I have four maps:
(1) obf for my region, downloaded with OsmAnd (about 100x100km)
(2) osm downloaded via JOSM, covering only the area of my changes (10x5km)
(3) osm with my changes to (2), saved in JOSM (10x5km)
(4) obf created by OsmMapCreator with (3) as input (10x5km)
Which ones if any of these are the "old" and "new" files? 

"... use utility from OsmAndMapCreator to generate "diff" file...":
Can you give more detail about "diff file" creation? I didn't find any information about this.

Thanks a lot again.

**Reply by vshcherb** [reply]
https://github.com/osmandapp/OsmAnd-tools/blob/master/java-tools/OsmAndMapCreatorUtilities/src/main/java/net/osmand/MainUtilities.java#L386 - here is a small description it requires 2 obf files which could be generated with https://github.com/osmandapp/OsmAnd-tools/blob/master/java-tools/OsmAndMapCreatorUtilities/src/main/java/net/osmand/MainUtilities.java#L377

**Reply by wega789** [reply]
This helped me very much. I did some tests and this is exactly what I wanted. What a great software!

I digged a bit into the code snippet and had a look at the batch files included in OsmAndMapCreator. Finally I wrote a Windows 10 batch file to simplify my intended workflow with drag & drop instead of handling many pathnames and executing several OsmMapCreator utility commands. Maybe it is helpful for others:

`@ECHO OFF
REM installDir: extraction directory of OsmAndMapCreator-main.zip, distrMap: name of officially distributed obf. Both must not be in quotes!
SET installDir=(your path)\OsmAndMapCreator-main
SET distrMap=Germany_baden-wuerttemberg_tubingen_europe.obf

SET oldname=%~dpn1%
SET oldosm="%oldname:~0,-2%.osm"
SET newosm=%1
SET oldobf="%oldname:~0,-2%.obf"
SET newobf="%oldname%.obf"
SET diffobf="%distrMap:\~0,-4%_diff%oldname:\~-1%.obf"
java.exe -Xms64M -Xmx512M -cp "%installDir%\OsmAndMapCreator.jar;%installDir%\lib\*.jar" net.osmand.MainUtilities generate-obf %oldosm%
java.exe -Xms64M -Xmx512M -cp "%installDir%\OsmAndMapCreator.jar;%installDir%lib\*.jar" net.osmand.MainUtilities generate-obf %newosm%
java.exe -Xms64M -Xmx512M -cp "%installDir%\OsmAndMapCreator.jar;%installDir%lib\*.jar" net.osmand.MainUtilities generate-obf-diff %oldobf% %newobf% %diffobf%
ECHO ----------
ECHO %diffobf% created, move it to phone!
PAUSE`

Explanation:
Path to extraction folder of OsmAndMapCreator must be set in "installDir". "distrMap" is the filename (without path) of the OsmAnd region map which I am going to edit. On my Android device it can be found in folder "...\Android\data\net.osmand.plus\files" (e.g. "Germany_baden-wuerttemberg_tubingen_europe.obf", different from the displayed map name in OsmAnd maps menu!).

Of course one has to use an OpenStreetMap editor, in my case JOSM. Since I only want personal modifications in relatively small areas I downloaded such areas of interest from OpenStreetMap. Using a smaller area greatly reduces download and processing time. Before applying any changes the downloaded data are saved with an arbitrary file name, e.g. "Tuebingen.osm". Then the map is modified and saved again in the same (!) directory. One must append an underscore and one character to the filename (e.g. "Tuebingen_1.osm"). The character is intended as a version number.

Now everything is ready for creating a final obf file which describes the differences. This is done by simply dragging the modified (!) file (here "Tuebingen_1.osm") and dropping it on the .bat file. This opens a console window, calls OsmMapCreator three times and is busy for a while. Finally six new files are created in the directory of the osm's. The important one is a .obf with the same name as distrMap and appended version information (here "Germany_baden-wuerttemberg_tubingen_europe_diff1.obf"). This file has to be copied into the above-mentioned folder on the Android device. And voila, the changes are visible. It sounds more complicated than it is.

So far it worked for me in moving, deleting, adding ways, changing their appearance and moving points. I did not test relations or routing. I noticed that the used executed utility commands overwrite existing files without warning. The batch file is not at all resistant against violation of the filename conventions above. And the workflow relies on the observation, that OpenStreetMap-IDs of modified items do not change between generation of the OsmAnd offline map and the download of the local area of interest in JOSM. I don't know if this will still be valid in the future.

Thanks again to vshcherb.
